builtin/providers/terraform/data_source_state.go

```diff
@@ -76,67 +29,83 @@ func dataSourceRemoteStateRead(d *schema.ResourceData, meta interface{}) error {
 	log.Printf("[DEBUG] Initializing remote state backend: %s", backendType)
 	f := backendinit.Backend(backendType)
 	if f == nil {
-		return fmt.Errorf("Unknown backend type: %s", backendType)
+		return cty.NilVal, diags.Append(tfdiags.AttributeValue(
+			tfdiags.Error,
+			fmt.Sprintf("Unknown backend type: %s", backendType),
+			"",
+			cty.Path(nil).GetAttr("backend"),
+		))
 	}
 	b := f()
 
-	schema := b.ConfigSchema()
-	rawConfig := d.Get("config")
-	configVal := hcl2shim.HCL2ValueFromConfigValue(rawConfig)
+	config := d.GetAttr("config")
+	newState["config"] = config
 
+	schema := b.ConfigSchema()
 	// Try to coerce the provided value into the desired configuration type.
-	configVal, err := schema.CoerceValue(configVal)
+	configVal, err := schema.CoerceValue(config)
 	if err != nil {
-		return fmt.Errorf("invalid %s backend configuration: %s", backendType, tfdiags.FormatError(err))
+		return cty.NilVal, diags.Append(tfdiags.AttributeValue(
+			tfdiags.Error,
+			fmt.Sprintf("invalid %s backend configuration: %s", backendType, tfdiags.FormatError(err)),
+			"",
+			cty.Path(nil).GetAttr("backend"),
+		))
 	}
+
 	validateDiags := b.ValidateConfig(configVal)
 	if validateDiags.HasErrors() {
-		return validateDiags.Err()
+		return cty.NilVal, diags.Append(validateDiags.Err())
 	}
 	configureDiags := b.Configure(configVal)
 	if configureDiags.HasErrors() {
-		return configureDiags.Err()
+		return cty.NilVal, diags.Append(configureDiags.Err())
 	}
 
-	// environment is deprecated in favour of workspace.
-	// If both keys are set workspace should win.
-	name := d.Get("environment").(string)
-	if ws, ok := d.GetOk("workspace"); ok && ws != backend.DefaultStateName {
-		name = ws.(string)
+	var name string
+	if d.Type().HasAttribute("workspace") {
```

Comment:
Thanks Martin, the concept makes good sense. 
Can you elaborate on how I would get the resource's schema for the tests? I'm either missing some context or tacking this too early (or both!)

Reply:
In one of the comments above I'd suggested moving the resource type schema out into its own function so that it can live in the same file as the resource type implementation. A secondary benefit of that would be that you could then call the function directly from the tests to get the schema here.